contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,479 @@
+service DevicePlugin {
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised
+	string resourceName = 3;
+}
+
+message RegisterResponse {
+	// Wether the request succeded or not
+	bool success = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process
+	string error = 2;
+}
+
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices sent
+message AllocateRequest {
+	repeated string deviceIDs = 1;
+}
+
+message AllocateResponse {
+	repeated DeviceRuntimeSpec spec = 1;
+}
+
+// ListWatch returns a stream of List of Devices
+// Whenever a Device state change or a Device disapears, ListWatch
+// returns the new list
+message ListWatchResponse {
+	repeated Device devices = 1;
+}
+
+// The list to be added to the CRI spec
+message DeviceRuntimeSpec {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
+	// Mounts for the container.
+	repeated Mount mounts = 2;
+}
+
+// E.g:
+// struct Device {
+//    ID: "GPU-fef8089b-4820-abfc-e83e-94318197576e",
+//    State: "Healthy",
+//}
+message Device {
+	string id = 1;
+	string state = 2;
+
+	DeviceRuntimeSpec spec = 3;
+}
+```
+
+We are expecting the State property to have some well known values:
+```
+const (
+	// Healthy means that the device is healty
+	Healty = "Healthy"
+	// UnHealthy means that the device is unhealty
+	Unhealthy = "Unhealty"
+)
+```
+
+### HealthCheck and Failure Recovery
+
+We want Kubelet as well as the Device Plugins to recover from failures
+that may happen on any side of this protocol.
+
+At the communication level, gRPC is a very strong piece of software and
+is able to ensure that if failure happens it will try it's best to recover
+through exponential backoff reconnection and Keep Alive checks.
+
+The proposed mechanism intends to replace any device specific handling in
+Kubelet. Therefore in general, device plugin failure or upgrade means that
+Kubelet is not able to accept any pod requesting a Device until the upgrade
+or failure finishes.
+
+If a device fails, the Device Plugin should signal that through the
+`ListAndWatch` gRPC stream. We then expect Kubelet to fail the Pod.
+
+If any Device Plugin fails the behavior we expect depends on the task Kubelet
```

Comment:
let's imagine i have 100 nodes, each node has 4 gpus.

if 50 of the gpus fail or go unhealthy, where do we expect to make that visible up to the cluster operator?

we discuss removing failed device plugins from the resource advertised by node status which i assume is removing it from allocatable, but the device still remains in capacity?  so if i wanted to know which nodes had potential device failures i have to monitor where device capacity != allocatable?  do we expect the kubelet to log some audit trail for that information?  do we want auditing of health transitions to be the domain of the plugin?